Documentation explaining the process for general secret rotation (#374)
SUSE/kubecf: doc/secret_rotation_general.md

# Secret Rotation

> Note, this document explains general rotation of secrets.

Reviewer: ```suggestion
Note, this document explains the general rotation of secrets.
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

The instructions to rotate the CCDB keys specifically are in
[a separate document](secret_rotation.md).

> The audience of this document are

Reviewer: ```suggestion
The audience of this document is:
```

Author: I am partial to the `are`, because what follows is all plural.

Reviewer: `is` refers to `the audience`, therefore the only right form of the verb to be is `is` here.

  - Developers working on KubeCF.

  - Operators deploying KubeCF.

# Background

> One of the features KubeCF (or rather the CF operator it sits on top of)

Reviewer: ```suggestion
One of the features KubeCF (or rather the cf-operator it sits on top of)
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

> provides is the ability to declare secrets (passwords, certificates, ...)

Reviewer: ```suggestion
provides is the ability to declare secrets (passwords and certificates)
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

and have the system automatically generate something suitably random
for such on deployment, and distribute the results to the pods using them.

This removes the burden from human operators to come up with lots of
such just to have all the internal components of KubeCF properly wired
up for secure communication.

> However, even with this operators may wish to change such secrets from

Reviewer: ```suggestion
However, even with this, operators may wish to change such secrets from
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

time to time, or on a schedule. In other words, re-randomize the
board, and limit the lifetime of any particular secret.

As a note on terminology, this kind of change is called
__rotating a secret__.

This document describes how this can be done, in the context of KubeCF.

# Finding secrets

Retrieve the list of all secrets maintained by a KubeCF deployment via

>     kubectl get quarkssecret -n kubecf

Reviewer: ```suggestion
    kubectl get quarkssecret --namespace kubecf
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

To see the information about a specific secret, for example the NATS password, use

>     kubectl get quarkssecret -n kubecf kubecf.var-nats-password -o yaml

Reviewer: ```suggestion
    kubectl get quarkssecret --namespace kubecf kubecf.var-nats-password --output yaml
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

Note that each quarkssecret has a corresponding regulare k8s secret it
controls.

>     kubectl get secret -n kubecf

Reviewer: ```suggestion
    kubectl get secret --namespace kubecf
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

>     kubectl get secret -n kubecf kubecf.var-nats-password -o yaml

Reviewer: ```suggestion
    kubectl get secret --namespace kubecf kubecf.var-nats-password --output yaml
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

# Requesting a rotation for a specific secret

We keep using `kubecf.var-nats-password` as our example secret.

To rotate this secret:

>   1. Create yaml file for a ConfigMap of the form:

Reviewer: ```suggestion
  1. Create a YAML file for a ConfigMap of the form:
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

         ---				   
         apiVersion: v1			   
         kind: ConfigMap			   
         metadata:			   
           name: rotate-kubecf.var-nats-password
           labels:			   
             quarks.cloudfoundry.org/secret-rotation: "true"
         data:				   
           secrets: '["kubecf.var-nats-password"]'

     Note, while the name of this ConfigMap can be technically

>      anything (allowed by k8s syntax) I personally prefer to use a

Reviewer: Instead of making the document personal ("I personally"), prefer using a collective terminology. E.g.
`..., we recommend using...`

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

     name derived from the name of the secret itself, to make the
     connection clear.

>      Note further that while this example rotates only a single secret

Reviewer: ```suggestion
     Note further that while this example rotates only a single secret,
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

     the `data.secrets` key accepts an array of secret names, allowing
     the simultaneous rotation of many secrets together.

>   2. Apply this ConfigMap using

Reviewer: ```suggestion
  2. Apply this ConfigMap using:
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

>          kubectl apply -n kubecf -f /path/to/your/yaml/file

Reviewer: ```suggestion
         kubectl apply --namespace kubecf -f /path/to/your/yaml/file
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

>   3. The CF operator will see that this ConfigMap is for it to

Reviewer: ```suggestion
  3. The cf-operator will process this ConfigMap due to the label `quarks.cloudfoundry.org/secret-rotation: "true"`
```

Reviewer: This line is probably too long, you will need to adjust with your editor to also play well with the following lines.

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

>      process, because of the special label

Reviewer: ```suggestion
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

> 

Reviewer: ```suggestion
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

>          quarks.cloudfoundry.org/secret-rotation: "true"

Reviewer: ```suggestion
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

> 

Reviewer: ```suggestion
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

     and knows that it has to invoke a rotation of the referenced
     secrets.

>      The actions of the operator can be followed in its log.

Reviewer: ```suggestion
     The actions of the cf-operator can be followed in its log.
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

>    4. After the operator has done the rotation, i.e. has not only

Reviewer: ```suggestion
   4. After the cf-operator has done the rotation, i.e. has not only
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980

      changed the secrets, but also restarted all affected pods (the
      users of the rotated secrets), delete the trigger config map
      again:
      

>          kubectl delete -n kubecf -f /path/to/your/yaml/file

Reviewer: ```suggestion
         kubectl delete --namespace kubecf --filename /path/to/your/yaml/file
```

Author: Addressed with commit 5d2c9c3009c76e029cbf5ce26e164e6bd5cd0980
